### ISSUE TITLE
Conditional fragments does not work as expected for Interfaces

### ISSUE THREAD

**Issue by epii-1** [LiUGraphQL/woo.sh#103]
While trying to find the cause of the issues I ran into while trying to implement, and test, conditional fragments for unions, I did similar tests for interfaces. These, very simple, tests gave similar results to those of unions, i.e. the conditioning is not working as expected.

I modified the `droid`-type in the test schema to contain the field `bestFriend: Character`, then created two droids with

``` graphQL
mutation {
                createDroid(data: {
                    name: "test"
                    bestFriend: { createDroid: {name: "test2"
                                                              primaryFunction: "something" }}
                    }) { id }}
```

and then queried 

``` graphQL
query {
            droid(id : someid){
                name
                bestFriend {
                    name
                    ... on Droid {
                        primaryFunction
                    } 
         }}}
```

which works just fine and as expected. However, both 
``` graphQL
query {
            droid(id : someid){
                name
                bestFriend {
                    name
                    ... on Human {
                        totalCredits
                    }
         }}}
```
and 
``` graphQL
query {
            droid(id : someid){
                name
                bestFriend {
                    name
                    ... on Droid {
                        primaryFunction
                    } 
                    ... on Human {
                        totalCredits
                    }
         }}}
```
throws 
```
Missing field totalCredits in {
  "name": "test2",
  "primaryFunction": "something",
  "__typename": "Droid"
}
```
and simply returns `null` for the whole query.

It feels like there is no conditioning being used here at all, and as if the resolver tries to resolve **all** requested fields instead of just the ones requested for the matching type.
I am wondering if this might be a pre v3 vs post v3 problem with Apollo, as they seem to have changed the behavior of fragments in general with the release of v3. There also seems to be some pre v3 reports of fragments misbehaving in Apollo.

@hartig @keski @rcapshaw

**Comment by keski** [comment]
I can't seem to replicate this exception. Using the original `friends` link I did:
```graphql
mutation {
  createDroid(
    data: {
      name: "C3PO"
      appearsIn: EMPIRE
      friends: [
        { createDroid: { name: "R2D2", appearsIn: EMPIRE, primaryFunction:"REPAIR" } }
        { createHuman: { name: "R2D2", appearsIn: EMPIRE, totalCredits:1000 } }
      ]
    }
  ) {
    id
    name
    friends {
      __typename
      ... on Human {
        totalCredits
      }
      ...on Droid {
        primaryFunction
      }
    }
  }
}
```
This worked as expected.

I then tried the following queries:
```graphql
query {
  droid(id: "Droid/12345") {
    name
    friends {
      name
      ... on Human {
        totalCredits
      }
    }
  }
}

query {
  droid(id: "Droid/12345") {
    name
    friends {
      name
      ... on Droid {
        primaryFunction
      }
      ... on Human {
        totalCredits
      }
    }
  }
}
```

This too behaved as expected.

If you can replicate it after dropping the DB, could you also post your updated schema (`bestFriend` is not in the DB schema in the repo)?

@epii-1

**Comment by epii-1** [comment]
@keski Want to bet we are using different versions of something (probably either apollo or arango)?
As you query/mutation does not work for me.

If I remove
``` graphQL
...
      ... on Human {
        totalCredits
      }
      ...on Droid {
        primaryFunction
      }
...
```
it works just fine. But with either of the fragments (or both) it does not work for me.

I just get the error:
```
graphQLErrors: [
    {
      message: "Cannot read property 'value' of undefined",
      locations: [Array],
      path: [Array],
      extensions: [Object]
    },
    { message: 'resolvedResult is not defined', extensions: [Object] }
  ],
  networkError: null,
  message: "GraphQL error: Cannot read property 'value' of undefined\n" +
    'GraphQL error: resolvedResult is not defined',
  extraInfo: undefined
```

As for the `bestFriend`-part: I just added the field `bestFriend: Character` to the droid-type.

With that being said: If it works for you, I suspect we can just close this issue and merge 97 > 102.
(Please just double check that you don't get the same result I did for what you suggested in https://github.com/LiUGraphQL/woo.sh/pull/97#issuecomment-674743412)

**Comment by keski** [comment]
Could you try rebuilding the `graphql-example-server/` from scratch? Since the npm dependencies are installed when the `woo.sh` script is run we should be running the exact same versions. It would be great if we could figure out what's happening here.

**Comment by epii-1** [comment]
Deleted the whole folder and rebuilt. Still same problem.

**Comment by epii-1** [comment]
The original issue here was actually **not** solved with #109. But we managed to find the real root cause: It's an issue with how I ran my tests. As I ran them similar to how `client-tests` are run instead of by using the playground, I had to install `apollo-cache-inmemory`, `apollo-client` and `apollo-server-testing`. One of these (probably client) seems to change the behavior of my queries, especially the fragments part. When running the same queries through the playground it works just fine and returns the expected results.

In short this should only be an issue when trying to automate tests for fragments, and otherwise this turned out to be a non-issue.

**Comment by keski** [comment]
@epii-1 I'll have a look at it. Which tests are you referring too? Could you provide a link to the code?

**Comment by epii-1** [comment]
Robin, I'm referring to the original tests of this issue, ran in a similar manner to how `client-tests.js` was run earlier (you saw me do it earlier while we noticed issue #109 ). You should be able to replicate it by using what you did in https://github.com/LiUGraphQL/woo.sh/issues/103#issuecomment-678995062, but by running the mutation and query from file instead of in the playground.
